aspnetcore/fundamentals/dependency-injection.md

```diff
@@ -133,94 +139,93 @@ public void Configure(IApplicationBuilder app, IOptions<MyOptions> options)
+* [Factory-based middleware](<xref:fundamentals/middleware/extensibility>). <xref:Microsoft.AspNetCore.Builder.UseMiddlewareExtensions.UseMiddleware*> extension methods check if a middleware's registered type implements <xref:Microsoft.AspNetCore.Http.IMiddleware>. If it does, the <xref:Microsoft.AspNetCore.Http.IMiddlewareFactory> instance registered in the container is used to resolve the <xref:Microsoft.AspNetCore.Http.IMiddleware> implementation instead of using the convention-based middleware activation logic. The middleware is registered as a scoped or transient service in the app's service container.
+
+For more information, see <xref:fundamentals/middleware/write#per-request-middleware-dependencies>.
+
+Do ***not*** resolve a scoped service from a singleton. It may cause the service to have incorrect state when processing subsequent requests. It's fine to:
+
+* Resolve a singleton service from a scoped or transient service.
+* Resolve a scoped service from another scoped or transient service.
+
+By default, in the development environment, resolving a service from another service with longer lifetime throws an exception. For more information, see [Scope validation](#sv).
 
 ### Singleton
 
-Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created the first time they're requested (or when `Startup.ConfigureServices` is run and an instance is specified with the service registration). Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide user code to manage the object's lifetime in the class.
+Singleton lifetime services (<xref:Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddSingleton*>) are created either:
 
-In apps that process requests, singleton services are disposed when the <xref:Microsoft.Extensions.DependencyInjection.ServiceProvider> is disposed at app shutdown.
+* The first time they're requested.
+* When `Startup.ConfigureServices` is run and an instance is specified with the service registration.
+
+Every subsequent request uses the same instance. If the app requires singleton behavior, allowing the service container to manage the service's lifetime is recommended. Don't implement the singleton design pattern and provide code to dispose of the singleton. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.
```

Comment:
Option 2:

```suggestion
Every subsequent request uses the same instance. If the app requires singleton behavior, allow the service container to manage the service's lifetime. Don't implement the singleton design pattern to manage the lifetime of the service. Services should never be disposed by code that resolved the service from a container. If a type or factory is registered as a singleton, the container will dispose the singleton automatically.
```

This is a separate suggestion because the _user code_ thing has been discussed previously and I don't want to bundle it all together. It also includes the preceding suggestion.